98 rides have link to next ride

clu723 opened the pull request:
Changes:
- Ride model:
  - belongs_to :next_ride, class_name: "Ride", optional: true
  - has_one :previous_ride, class_name: "Ride", foreign_key: "next_ride_id", dependent: :destroy
- Ride Schema:
  - t.references :next_ride, foreign_key: { to_table: :rides }
    - {to_table: :rides} tells Rails, "This is a foreign key to the rides table" (not to a table named previous_rides, which doesn’t exist). 
- A multi-stop/roundtrip ride will be shown as one ride in the rides table, but each sub-ride to a destination will be counted as a ride in the 5310 report generation.
- Added "Multiple Stops" column to rides table to indicate if a ride has multiple stops, updates by itself
- removed obsolete rides filter route, view, action method
- Can add/remove stops in new and edit rides view (app/javascript/add_stop.js)
- Updated app/javascript/autocomplete.js to be able to autocomplete new stop fields that are added by user
- Updated rides create, update, destroy controller actions to handle multi-stop rides
- removed obsolete rides table columns
- Added Notes About Location field to show view of rides (for drivers to see)
- Shift show view and Driver's today view only shows the head ride if ride is multi-stop
- Rides Edit and Show views' back buttons now redirect to previous page instead of a predetermined page
- Origin address defaults to passenger's home address

Issues:
- cucumber tests (not sure if it's worth spending our remaining time on this)

Unrelated Issues Found (Fixed in #133 ):
- Notes about location and Address Name should be in address schema rather in rides - also to be autofilled 
  - should also be in edit views
- emailed driver is binary type but we're importing "sent" into it so it currently on displays "No" in datatables

ZhixiaLin commented:
<img width="987" alt="image" src="https://github.com/user-attachments/assets/f0bd37cd-de0b-4f39-872b-6eb85dd60852" />

Would replacing "Info" with "Edit" be the right fix to resolve the cucumber failure?

clu723 commented:
The accessibility info comes from the Passenger data. I don't think it's necessary for them to be able to edit those fields in rides rather than in Passengers, since we're retrieving that information from Passengers when creating new rides.

AthenaJenkins commented:
Something to change:
It would be good to avoid having the browser try to automatically prompt with the browser's saved addresses.
In Chrome (and I would assume some other browsers too) it detects the "Street" field name and automatically pops up this little window:
![image](https://github.com/user-attachments/assets/4c2c2c5a-300e-4608-bb4f-58117842d977)
This isn't what we want though, as we have our own autocomplete functionality and this just gets in the way.

The line I added to fix this was `$("#ride_dest_address_attributes_street").attr("autocomplete", "ride-address");`.
But with the new multiple stops it will need to be readded and adjusted to work when a new stop is added.

Here's a little explanation for why this is the solution for this:

Historically, this could be avoided by setting the `autocomplete` tag to `"off"`. However, chrome (and others) now ignore any `autocomplete="off"` tags, and so instead you have to set `autocomplete` to some custom name, a suggestion I got from this post: https://stackoverflow.com/a/30976223.

I originally tried to do this in the new ride view, simply adding a `autocomplete="ride-address"` tag. However, very annoyingly, the jQuery autocomplete module actually will set the autocomplete tag to off for you, which while in the past would have been convenient it now just overrides the value we set. And so we have to override this again in the javascript after the jQuery autocomplete is set up. 

I probably should have added more documentation of this somewhere so it would have been clearer, as it's sort of a weird line, but that's the explanation.

ZhixiaLin commented:
@wilsonylai @clu723 
After merging branch 102 and updating branch 98, I discovered during testing that some parts of the code were still not fully synchronized with the main branch, even though GitHub did not show any additional error messages or prompt for branch updates.
I initially tried using git merge to resolve the discrepancies, but it did not update anything else. I then tried git rebase, which helps ensure a cleaner branch history. However, during the rebase process, I encountered several conflicts, some of which involved files or features that I was not directly involved with. As a result, there may be areas where the conflict resolution could benefit from further review.
Although GitHub indicates that this branch can be automatically merged, please keep an eye on this issue during the review, and feel free to reach out to relevant team members if anything looks uncertain. Thanks!

clu723 commented:
@ZhixiaLin Everything seems to be good except that the Driver's Today view had two wheelchair columns. Fixed that now. Thank you for merging!

XZhaoF reviewed (changes requested):
In the edit page for rides, you can delete all the stops aside from the origin address,

clu723 commented:
@XZhaoF Wait why?

XZhaoF commented:
> @XZhaoF Wait why?

Like it's a bug, we shouldn't be able to do that yet we can

clu723 commented:
@XZhaoF Ohhh ok sorry I misunderstood. It's fixed now.

AthenaJenkins reviewed (commented):
Seems to work well overall.
Some notes relating to the actual functionality:
- The shift show page (where the dispatchers will be editing the rides the most) doesn't show any of the updated information, including the stuff about if a ride has multiple stops.
- There currently, as far as I can tell, there is no way for the dispatchers to actually see the additional stops except through the edit page (the only place we link the show page for a ride is from the today view for a driver). I feel like ideally we would able to see it from the index view somehow, but given that that is harder to implement I think there should at least be some way to go to the show page from the index view
 
Some notes about the code (we are in a rush so we might not get to these):
- Should `.byebug_history` be in our .gitignore? Is there any reason to track this file?
- I feel like some of the functionality currently put in the controller should be put into the model. I think it would be good to make the code to fetch all of the ride heads (`Ride.where.not(id: Ride.select(:next_ride_id).where.not(next_ride_id: nil))`) a model method (named something like `Ride.heads`) rather than repeating it in the controller multiple times. Similarly, I feel like we could move the destroy stuff currently in the controller method in a `before_destroy` callback.
- Editing a ride currently just destroys the ride and makes a new one. This is technically fine for now, but in the future if anything ever belongs_to a ride this will cause those relations to break, as the foreign key pointing to the ride will no longer accurately point to it. I'm not sure if we have time to fix this, but this is definitely a big enough concern to make sure its included in the wiki somewhere.
- Currently if editing or creating a ride fails, the website will send a 500 error instead of rerendering the form nicely. Not currently an issue as I don't think there is any way to fill out the form such that the ride creation fails, but it's not ideal. Probably another thing for future students to clean up though
- At the top of the new ride view there is `<%= javascript_importmap_tags %>`. I'm a little confused on why we need that, but it does break passenger autocomplete javascript if we remove it. What is that doing?
- In the new and _stop_fields ride views, we should probably use ActiveView form helpers instead of raw html

clu723 commented:
@AthenaJenkins 
- Updated shift view index page to contain "Multiple Stop" column
- Dispatchers can see all the destinations of the ride using the Edit button, not sure if we should add an additional "More Info" button if we already have that
- Removed byebug history file
- Removed the javascript import tag in new.html.erb and passenger autofill still works

ZhixiaLin commented on app/views/rides/new.html.erb:
```diff
@@ -1,199 +1,225 @@
+           </div>
+
+           <div class="mb-3">
+             <%= f.label :driver_id, "Driver" %>
```
For all `required: true` fields, I suggest add the `<span class="text-danger">*</span>` as a reminder
```suggestion
             <%= f.label :driver_id, "Driver" %>  <span class="text-danger">*</span>
```
also add this text-danger warning of `Emailed Driver` and `Passenger`, right now the `Van` has not been set `required: true`, I wasn't sure about this

clu723 replied:
I noticed this just earlier today, and I decided to implement this change in #140, giving an asterisk to Emailed Driver, Passenger, and Driver and making them required fields. I don't think Van needs to be assigned during ride creation, however.

DanielQLTan commented:
@clu723 Let's keep the button as "Last Day" for consistency with calendar page as well as UI alignment and same button sizes. If you are really against this wording, consider at least changing the language to "Prev. Day", and remember to update the calendar page as well.

clu723 commented:
@DanielQLTan Got it. Renamed to "Prev. Day" and "Prev. Month"
